cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/k8s/SparkContainerDriverLauncher.java

```diff
@@ -0,0 +1,258 @@
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+public class SparkContainerDriverLauncher {
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder())
+    .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+    .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec())
+    .create();
+  private static final String PROGRAM_JAR_NAME = "program.jar";
+  private static final String PROGRAM_JAR_EXPANDED_NAME = "program.jar.expanded.zip";
+  private static final String CDAP_APP_SPEC_KEY = "cdap.spark.app.spec";
+  private static final String DEFAULT_DELEGATE_CLASS = "org.apache.spark.deploy.SparkSubmit";
+  private static final String DELEGATE_CLASS_FLAG = "--delegate-class";
+
+  //TODO (CDAP-18315): following three lines need to be fixed once CDAP-18315 is resolved.
+  private static final String WORKING_DIRECTORY = "/opt/spark/work-dir/";
+  private static final String CCONF_PATH = WORKING_DIRECTORY + "cConf.xml";
+  private static final String HCONF_PATH = WORKING_DIRECTORY + "hConf.xml";
+
+  private static ArtifactFetcherService artifactFetcherService;
+
+  public static void main(String[] args) throws Exception {
+    //TODO (CDAP-18315): fix once CDAP-18315 is addressed.
+    setConfigurations(args);
+
+    String delegateClass = DEFAULT_DELEGATE_CLASS;
+    List<String> delegateArgs = new ArrayList<>();
+    for (int i = 0; i < args.length; i++) {
+      if (args[i].equals(DELEGATE_CLASS_FLAG)) {
+        delegateClass = args[i + 1];
+        i++;
+        continue;
+      }
+      delegateArgs.add(args[i]);
+    }
+
+    CConfiguration cConf = CConfiguration.create(new File(CCONF_PATH));
+    Configuration hConf = new Configuration();
+    hConf.addResource(new org.apache.hadoop.fs.Path("file:" + new File(HCONF_PATH).getAbsolutePath()));
+
+    ApplicationSpecification spec =
+      GSON.fromJson(hConf.getRaw(CDAP_APP_SPEC_KEY), ApplicationSpecification.class);
+    ArtifactLocalizerClient fetchArtifacts = createArtifactLocalizerClient(cConf, hConf);
+
+    //Create plugin location for storing plugin jars
+    Path pluginsLocation = new File(WORKING_DIRECTORY).getAbsoluteFile().toPath().resolve("artifacts_archive.jar")
+      .toAbsolutePath();
+    Files.createDirectories(pluginsLocation);
+
+    // Fetching plugin artifacts from app-fabric
+    for (Plugin plugin : spec.getPlugins().values()) {
+      File tempLocation = fetchArtifacts.localizeArtifact(plugin.getArtifactId());
+      String pluginName = String.format("%s-%s-%s.jar",
+                                        plugin.getArtifactId().getScope().toString(),
+                                        plugin.getArtifactId().getName(),
+                                        plugin.getArtifactId().getVersion().toString());
+      RunJar.unJar(tempLocation, pluginsLocation.resolve(pluginName).toFile(), RunJar.MATCH_ANY);
```

Comment:
For new code, please do it without Hadoop as dependency as much as possible. We have a BundleJarUtil for this purpose.

Reply:
replace this with BundleJarUtil too